components/service/glean/src/main/java/mozilla/components/service/glean/CounterMetricType.kt

```diff
@@ -0,0 +1,60 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.glean
+
+import mozilla.components.service.glean.storages.CountersStorageEngine
+import mozilla.components.support.base.log.logger.Logger
+
+/**
+ * This implements the developer facing API for recording counter metrics.
+ *
+ * Instances of this class type are automatically generated by the parsers at build time,
+ * allowing developers to record values that were previously registered in the metrics.yaml file.
+ *
+ * The counter API only exposes the [add] method, which takes care of validating the input
+ * data and making sure that limits are enforced.
+ */
+data class CounterMetricType(
+    override val disabled: Boolean,
+    override val category: String,
+    override val lifetime: Lifetime,
+    override val name: String,
+    override val sendInPings: List<String>
+) : CommonMetricData {
+
+    override val defaultStorageDestinations: List<String> = listOf("metrics")
+
+    private val logger = Logger("glean/CounterMetricType")
+
+    /**
+     * Add to counter value.
+     *
+     * @param amount This is the amount to increment the counter by, defaulting to 1 if called
+     * without parameters.
+     */
+    fun add(amount: Int = 1) {
+        // TODO report errors through other special metrics handled by the SDK. See bug 1499761.
+
+        if (!shouldRecord(logger)) {
+            return
+        }
+
+        if (amount <= 0) {
+            return
+        }
+
+        // Retrieve existing value
+        val snapshot = CountersStorageEngine.getSnapshot(storeName = sendInPings[0], clearStore = false)
```

Comment:
I like the idea of using a lambda much better.  I had initially just made this change to test and see if it would work and grab the right data but I wasn't happy with it and was looking for other ways to go about it.

Reply:
If you look above for my conversation with @mdboom regarding whether or not the values stored in user prefs would be different for each of the 'sendInPings' list entries, I was also concerned about trying to do so much work in the metric determining what values where stored and if they were going to be the same or could be different.  Moving this like this where the engine can perform the lambda operation for each store is a potential solution for that question as well.

Reply:
Perfect! #1527 was just merged so you should now be able to rebase ;)